### PR TITLE
[api][web][#393]Education Programs/ Offerings Enhancements Dev Part 2

### PR DESCRIPTION
**Pull request by guru-aot**
#393 

Form changes for the below 3 files will be taken care by @dheepak-aot, as he is working on the same form already for the story #722

image

Please ignore the code smells, as its the repeat of the same code in the dto layer. I dont know to clear it, if somebody has suggestions, please share.

**Institution Location- Create New Program Page**
Text Change:
- [x] "Is the Minimum age 19 and up?" _(Y/N)_ 
- [x] If No, Change the wording of the banner-_Reveal the Banner,  "This program is ineligible for StudentAid BC funding", Students need to be 19 and up.  If you require an appeal, please continue and complete this application. StudentAidBC will review your application" Set Program Status to 'Pending'_

image

** Institution Location- Add Offering Page**
- [x] Hide the Feild- "No Offering Dates" (For Continuous Entry)" _(Added to the Icebox for future Solution-ing)_ Hardcode to False
- [x] Hide the Feild- - "No Fixed Costs" (Added to the Icebox for future Solution-ing)_ Hardcode to False
---No Database change required. Columns remain either nullable or not-nullable

image

**Student Application- Select your Program Page**
- [x] When the Student Selects the Program from the Dropdown list in "The Program I will be Attending", _Provide The details of the Program Name, Description, Credential, and Delivery Method_

image

### CODE REVIEW

File: sources/packages/api/src/route-controllers/student/student.controller.ts [hunk 1]
```diff
@@ -146,6 +149,27 @@ export class StudentController extends BaseController {
     );
   }
 
+  /**
+   * This returns only a part of the EducationProgram details to fetch
+   * for the of Student
+   * @param programId
+   * @returns
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/student/student.controller.ts:156]: you missed this comment

File: sources/packages/api/src/route-controllers/student/student.controller.ts [hunk 2]
```diff
@@ -146,6 +149,27 @@ export class StudentController extends BaseController {
...
+   */
+  @Get("/education-program/:programId/summary")
+  async get(
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/student/student.controller.ts:159]: can you rename the function

File: sources/packages/api/src/services/education-program/education-program.service.ts [hunk 3]
```diff
@@ -40,6 +41,46 @@ export class EducationProgramService extends RecordDataModelService<EducationPro
...
+        "programs.name",
+        "programs.description",
+        "programs.credentialType",
+        "programs.credentialTypeOther",
+        "programs.deliveredOnSite",
+        "programs.deliveredOnline",
+      ])
+      .where("programs.id = :programId", { programId })
+      .getOne();
+
```
Comment (ann-aot) [inline sources/packages/api/src/services/education-program/education-program.service.ts:67]: IMO, the query should be an individual function that returns the repo as promise, and the DTO response can be created inside the controller or as a sep function, then only we could reuse the query right? if the service is specifically for a DTO, then we cant reuse it right? thoughts @guru-aot  @andrewsignori-aot  @dheepak-aot

File: sources/packages/web/src/composables/useFormioDropdownLoader.ts [hunk 4]
```diff
@@ -67,6 +67,30 @@ export function useFormioDropdownLoader() {
...
+    programId: number,
+    fieldId: string,
+  ) => {
+    const valueToBeLoaded = await EducationProgramService.shared.getStudentEducationProgram(
+      programId,
+    );
+    formioUtils.setComponentValue(
+      form,
+      fieldId,
+      `<h3><span style="background-color:transparent;color:#2965c5;"><strong>${valueToBeLoaded.name}</strong></span></h3>${valueToBeLoaded.description}
```
Comment (ann-aot) [inline sources/packages/web/src/composables/useFormioDropdownLoader.ts:82]: instead of inline style, you can create a class and use

File: sources/packages/web/src/composables/useFormioDropdownLoader.ts [hunk 5]
```diff
@@ -67,6 +67,30 @@ export function useFormioDropdownLoader() {
...
+      <table style="width:100%">
```
Comment (ann-aot) [inline sources/packages/web/src/composables/useFormioDropdownLoader.ts:83]: why `<table>` ?

Comment (guru-aot) [reply sources/packages/web/src/composables/useFormioDropdownLoader.ts:83]: Without table, it looks weird and we had to use &nbsp for the spaces, this is to be in sync with the wireframe. If you have suggestions, please tell.

File: sources/packages/api/src/route-controllers/student/models/student.dto.ts [hunk 6]
```diff
@@ -52,3 +52,16 @@ export interface FileCreateDto {
   size: number;
   mimetype: string;
 }
+
+export class StudentEducationProgramDto {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/student/models/student.dto.ts:56]: Is this DTO used to send data to the API?
Using a class and the validators have effect only to submit data to the API. If this is not the case, please convert it to an interface and remove the validators.

File: sources/packages/api/src/services/education-program/education-program.service.ts (hunk 3 above)
Comment (guru-aot) [reply sources/packages/api/src/services/education-program/education-program.service.ts:67]: I feel, we should restrict the call in the service level as the student is not authorized to see other columns like the institution. Else we would have used the getInstitutionProgram method of the same service and would have created a separate controller reusing the service.

File: sources/packages/api/src/route-controllers/student/student.controller.ts (hunk 2 above)
Comment (guru-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.ts:159]: the complete route looks like api/students/education-program/44/summary. I feel its explanatory, saying the education-program summary for a student.

File: sources/packages/web/src/composables/useFormioDropdownLoader.ts (hunk 5 above)
Comment (guru-aot) [reply sources/packages/web/src/composables/useFormioDropdownLoader.ts:83]: image

Comment (ann-aot) [reply sources/packages/web/src/composables/useFormioDropdownLoader.ts:83]: https://getbootstrap.com/docs/4.0/layout/grid/, you can use bootstrap classess

File: sources/packages/api/src/route-controllers/student/student.controller.ts (hunk 2 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.ts:158]: the is student controller, IMO, If the function name is `get` , then router should be `api/student/:id`

File: sources/packages/web/src/composables/useFormioDropdownLoader.ts (hunk 5 above)
Comment (guru-aot) [reply sources/packages/web/src/composables/useFormioDropdownLoader.ts:83]: So what's the difference?

File: sources/packages/api/src/services/education-program/education-program.service.ts [hunk 7]
```diff
@@ -40,6 +41,46 @@ export class EducationProgramService extends RecordDataModelService<EducationPro
...
+      .select([
+        "programs.id",
+        "programs.name",
+        "programs.description",
+        "programs.credentialType",
+        "programs.credentialTypeOther",
+        "programs.deliveredOnSite",
+        "programs.deliveredOnline",
+      ])
+      .where("programs.id = :programId", { programId })
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/services/education-program/education-program.service.ts:65]: I believe that we should restrict this a bit more. Can the code do the check if at least the program is approved?
```
.where("programs.approvalStatus = :approvalStatus", {
        approvalStatus: ApprovalStatus.approved,
      })
```

Comment (guru-aot) [reply sources/packages/api/src/services/education-program/education-program.service.ts:65]: I agree, even for the list we are showing only the approved programs

File: sources/packages/api/src/services/education-program/education-program.service.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/education-program/education-program.service.ts:67]: As @ann-aot mentioned, I believe that we has this conversation before and we had a meeting about it. Is there a reason to not follow it?

File: sources/packages/web/src/composables/useFormioDropdownLoader.ts (hunk 5 above)
Comment (ann-aot) [reply sources/packages/web/src/composables/useFormioDropdownLoader.ts:83]: IMO, it's not appropriate to use the table for these types of styles, in future if we want to add a custom styles or override an actual table, it will be difficult and break styles

File: sources/packages/api/src/route-controllers/student/student.controller.ts (hunk 2 above)
Comment (guru-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.ts:158]: the id of the student is not used in the query, its a plain query to fetch the education program description for any student irrespective. As its for the student, I have moved it to the student controller.

File: sources/packages/web/src/composables/useFormioDropdownLoader.ts (hunk 5 above)
Comment (guru-aot) [reply sources/packages/web/src/composables/useFormioDropdownLoader.ts:83]: OKay, let me change it.

File: sources/packages/api/src/route-controllers/student/student.controller.ts (hunk 2 above)
Comment (ann-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.ts:158]: Let me rephrase, IMO, it makes sense to give a function name as `get`, when the router is `api/student/:id`, here the case is different right?.  `get` is not an appropriate name for the function, in this scenario IMO. thoughts @andrewsignori-aot  @dheepak-aot

Comment (guru-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.ts:158]: Ho you meant the get :). I get it now 👍 Thanks will rename

File: sources/packages/api/src/services/education-program/education-program.service.ts (hunk 3 above)
Comment (guru-aot) [reply sources/packages/api/src/services/education-program/education-program.service.ts:70]: I might have missed it, I think I followed what we discussed. If I have missed something, please explain or we can have a call.

Comment (andrewsignori-aot) [reply sources/packages/api/src/services/education-program/education-program.service.ts:70]: Please follow what @ann-aot mentioned or organize a call and we can talk about it.

File: sources/packages/web/src/composables/useFormioDropdownLoader.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/composables/useFormioDropdownLoader.ts:83]: I do not see a reason to keep this HTML on Vue application. Can you please move this HTML to the form definition?

File: sources/packages/web/src/composables/useFormioDropdownLoader.ts [hunk 8]
```diff
@@ -67,6 +67,30 @@ export function useFormioDropdownLoader() {
     );
   };
 
+  // Get Program description for the selected program and set to the field (programDesc) in formio
+  const loadProgramDesc = async (
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/composables/useFormioDropdownLoader.ts:71]: This class is for methods related to loading dropdowns. Loading this information does not sound like something directly related to the dropdowns loading. Yes, the data is loaded in reaction to a dropdown event, but still not loading a data to a dropdown.

**Review by andrewsignori-aot — COMMENTED** [review]
Please connect the issue with the ticket.
Please take a look at the comments.
About the below comment in the description.
"_Please ignore the code smells, as its the repeat of the same code in the dto layer. I dont know to clear it, if somebody has suggestions, please share._"
SonarCloud is configured to ignore code duplications on ".dto" files. This expectation was added to SonarCloud with the intention of avoiding duplication warnings between API DTOs and Vue DTOs.
In the mentioned case, the duplication is API->API. Either way, renaming the class to something.dto will make SonarCloud ignore it.

File: sources/packages/api/src/route-controllers/student/student.controller.ts [hunk 9]
```diff
@@ -38,7 +40,11 @@ import { Readable } from "stream";
 import { defaultFileFilter, uploadLimits } from "../../utilities/upload-utils";
 import { StudentApplicationDTO } from "../application/models/application.model";
 import { Application } from "../../database/entities";
-import { determinePDStatus } from "../../utilities/student-utils";
+import {
+  determinePDStatus,
+  deliveryMethod,
+} from "../../utilities/student-utils";
+import { credentialTypeToDisplay } from "../../utilities/credential-type-utils";
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/student/student.controller.ts:47]: i think u missed to push this file

Comment (guru-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.ts:47]: you mean credentialTypeToDisplay? its already present. i just reused the utility

Comment (ann-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.ts:47]: okay 👍

**Review by abschwenker — APPROVED** [review]

File: sources/packages/web/src/composables/useFormioComponentLoader.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+import { useFormioUtils } from ".";
+
+/**
+ * Common methods to load components data on Form.IO that could
+ * be reusable or at least simplify the form data load logic.
+ * @returns Methods to help loading the data into Form.IO forms.
+ */
+export function useFormioComponentLoader() {
+  const formioUtils = useFormioUtils();
+  // get offering date of the selected offering and set to the hidden field (selectedOfferingDate) in formio
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/composables/useFormioComponentLoader.ts:12]: As agreed, please start the sentence with a capital letter and end it with a period.

File: sources/packages/api/src/route-controllers/student/student.controller.ts [hunk 11]
```diff
@@ -146,6 +153,32 @@ export class StudentController extends BaseController {
     );
   }
 
+  /**
+   * This returns only a part of the EducationProgram details for the student
+   * @param programId
+   * @returns
+   */
+  @Get("/education-program/:programId/summary")
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/student/student.controller.ts:161]: Do we need to have the "_summary_" appended here? I believe that this could be reused anytime that there is a need to retrieve the program data for a student.
Following this, the property "_credentialType_" on DTO could be renamed to "_credentialTypeToDisplay_" allowing the DTO in the future to return the real "_credentialType_" from the database.

**Review by dheepak-aot — APPROVED** [review]
LGTM. Please address the review comments.

File: sources/packages/api/src/route-controllers/student/student.controller.ts (hunk 11 above)
Comment (guru-aot) [reply sources/packages/api/src/route-controllers/student/student.controller.ts:161]: removed the /summary

**Review by andrewsignori-aot — APPROVED** [review]
Looks good 👍 
Thanks for doing the changes and adjusting the Form.IO definition to remove the custom HTML.
